lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp

```diff
@@ -781,24 +781,48 @@ SILCombiner::visitInjectEnumAddrInst(InjectEnumAddrInst *IEAI) {
 
   // Ok, we have a payload enum, make sure that we have a store previous to
   // us...
-  SILBasicBlock::iterator II = IEAI->getIterator();
+  auto allocStackOperand = IEAI->getOperand();
   StoreInst *SI = nullptr;
   InitEnumDataAddrInst *DataAddrInst = nullptr;
   ApplyInst *AI = nullptr;
   Operand *EnumInitOperand = nullptr;
-  for (;;) {
-    if (II == IEAI->getParent()->begin())
+  for (auto usersIt = allocStackOperand->use_begin(); usersIt != allocStackOperand->use_end(); ++usersIt) {
+    auto currUser = usersIt->getUser();
+    // Find a Store whose destination is taken from an init_enum_data_addr
+    // whose address is same allocation as our inject_enum_addr.
+    DataAddrInst = dyn_cast<InitEnumDataAddrInst>(currUser);
+    if (DataAddrInst) {
+      break;
+    }
+  }
+  if (DataAddrInst) {
+    // Found the store to this enum payload. Check if the store is the only use.
+    if (!DataAddrInst->hasOneUse())
       return nullptr;
-    --II;
-    SI = dyn_cast<StoreInst>(&*II);
+    SI = dyn_cast<StoreInst>(DataAddrInst->use_begin()->getUser());
     if (SI) {
-      // Find a Store whose destination is taken from an init_enum_data_addr
-      // whose address is same allocation as our inject_enum_addr.
-      DataAddrInst = dyn_cast<InitEnumDataAddrInst>(SI->getDest());
-      if (DataAddrInst && DataAddrInst->getOperand() == IEAI->getOperand())
-        break;
-      SI = nullptr;
+      assert(SI->getDest() == DataAddrInst);
```

Comment:
Also, I think you can eliminate this if statement by doing something like:

assert((!SI || (SI->getDest() == DataAddrInst)) && "...");